File: tree-medium/105.从前序与中序遍历序列构造二叉树.java
/*
 * @lc app=leetcode.cn id=105 lang=java
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// @lc code=start

// 202/202 cases passed (4 ms)
// Your runtime beats 49.33 % of java submissions
// Your memory usage beats 85.11 % of java submissions (38.3 MB)
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return build(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }

    TreeNode build(int[] preorder, int preStart, int preEnd, int[] inorder, int inStart, int inEnd) {
        if (preStart > preEnd) {
            return null;
        }

        int val = preorder[preStart];
        int idx = 0;
        for (int i = 0; i < inorder.length; i++) {
            if (inorder[i] == val) {
                idx = i;
                break;
            }
        }
        TreeNode root = new TreeNode(val);
        // 左子树的节点个数
        int leftSize = idx - inStart;
        root.left = build(preorder, preStart + 1, preStart + leftSize, inorder, inStart, idx - 1);
        root.right = build(preorder, preStart + leftSize + 1, preEnd, inorder, idx + 1, inEnd);
        return root;
    }
}
// @lc code=end

File: tree-medium/105.从前序与中序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=105 lang=cpp
 *
 * [105] 从前序与中序遍历序列构造二叉树
 */

/*
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include <unordered_map>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
   public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        for (int i = 0; i < inorder.size(); i++) {
            mem[inorder[i]] = i;
        }
        return dfs(preorder, inorder, 0, preorder.size() - 1, 0, inorder.size() - 1);
    }

   private:
    unordered_map<int, int> mem;
    TreeNode *dfs(vector<int> &preorder, vector<int> &inorder, int pre_start, int pre_end, int in_start, int in_end) {
        if (pre_start > pre_end) {
            return nullptr;
        }
        TreeNode *root = new TreeNode(preorder[pre_start]);
        int idx = mem[preorder[pre_start]];
        int size = idx - in_start;
        root->left = dfs(preorder, inorder, pre_start + 1, pre_start + size, in_start, idx - 1);
        root->right = dfs(preorder, inorder, pre_start + size + 1, pre_end, idx + 1, in_end);
        return root;
    }
};
// @lc code=end
